packages/typed-inject/README.md

```diff
@@ -0,0 +1,264 @@
+    static inject = tokens('bar');
+}
+const foo /*: Foo*/ injector.injectClass(Foo);
+```
+
+#### `injector.injectFunction(fn: InjectableFunction)`
+
+This methods injects the function with requested tokens and returns the return value of the function.
+When there are any problems in the dependency graph, it gives a compiler error.
+
+```ts
+function foo(bar: number) {
+    return bar + 1;
+}
+foo.inject = tokens('bar');
+const baz /*: number*/ injector.injectFunction(Foo);
+```
+
+#### `injector.resolve(token: Token): CorrespondingType<TContext, Token>`
+
+The `resolve` method lets you resolve tokens by hand. 
+
+```ts
+const foo = injector.resolve('foo');
+// Equivalent to:
+function retrieveFoo(foo: number){
+    return foo;
+}
+retrieveFoo.inject = tokens('foo');
+const foo2 = injector.injectFunction(retrieveFoo);
+```
+
+#### `injector.provideValue(token: Token, value: R): Injector<ChildContext<TContext, Token, R>>`
+
+Create a child injector that can provide value `value` for token `'token'`. The new child injector can resolve all tokens the parent injector can as well as `'token'`.
+
+```ts
+const fooInjector = injector.provideValue('foo', 42);
+```
+
+#### `injector.provideFactory(token: Token, factory: InjectableFunction<TContext>, scope = Scope.Singleton): Injector<ChildContext<TContext, Token, R>>`
+
+Create a child injector that can provide a value using `factory` for token `'token'`. The new child injector can resolve all tokens the parent injector can, as well as the new `'token'`.
+
+With `scope` you can decide whether the value must be cached after the factory is invoked once. Use `Scope.Singleton` to enable caching (default), or `Scope.Transient` to disable caching.
+
+```ts
+const fooInjector = injector.provideFactory('foo', () => 42);
+function loggerFactory(target: Function | undefined) {
+    return new Logger((target && target.name) || '');
+}
+loggerFactory.inject = tokens(TARGET_TOKEN);
+const fooBarInjector = fooInjector.provideFactory('logger', loggerFactory, Scope.Transient)
+```
+
+#### `injector.provideFactory(token: Token, Class: InjectableClass<TContext>, scope = Scope.Singleton): Injector<ChildContext<TContext, Token, R>>`
+
+Create a child injector that can provide a value using instances of `Class` for token `'token'`. The new child injector can resolve all tokens the parent injector can, as well as the new `'token'`.
+
+Scope is also supported here, for more info, see `provideFactory`.
+
+### `Scope`
+
+The `Scope` enum indicates the scope of a provided injectable (class or factory). Possible values: `Scope.Transient` (new injection per resolve) or `Scope.Singleton` (inject once, and reuse values). It generally defaults to `Singleton`. 
+
+### `tokens`
+
+The `tokens` function is a simple helper method that makes sure that an `inject` array is filled with a [tuple type filled with literal strings](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#rest-parameters-with-tuple-types).
```

Comment:
How about calling this what it is? A function to create a tuple filled with literal strings. I think something like `createStringLiteralTuple` would make more sense than `tokens`. What are your thoughts on this?

Reply:
I think `createStringLiteralTuple` is a far to long name. I understand where you're coming from, but functionally it is used to define _injection tokens_. In the future I'm thinking of expanding the functionality of `typed-inject` to also except Symbols (as well as Strings) using TS 2.7's [`unique symbol`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#unique-symbol) type. 

In other words, `tokens` will always be used to define tokens, but might not be used to define solely _string tuples_ in the future. 

Does that make sense?

Reply:
Sounds good!